### PR TITLE
Use Data Center GPU Manager (DCGM) to get metrics instead of NVML

### PR DESCRIPTION
**Pull request by CoderHam**
- TODO fix windows build for the same

### CODE REVIEW

File: src/core/metrics.cc [hunk 1]
```diff
@@ -156,205 +156,215 @@ Metrics::EnableGPUMetrics()
...
 #else
-  nvmlReturn_t nvmlerr = nvmlInit();
-  if (nvmlerr != NVML_SUCCESS) {
-    LOG_WARNING << "failed to initialize, GPU metrics will not be available: "
-                << nvmlErrorString(nvmlerr);
+  dcgmReturn_t dcgmerr = dcgmInit();
+  if (dcgmerr != DCGM_ST_OK) {
+    LOG_WARNING
+        << "failed to initialize DCGM, GPU metrics will not be available: "
+        << errorString(dcgmerr);
```
Comment (deadeyegoodwin) [inline src/core/metrics.cc:175]: Is this a dcgm-specific error string method?

File: src/core/metrics.cc [hunk 2]
```diff
@@ -156,205 +156,215 @@ Metrics::EnableGPUMetrics()
...
-      LOG_WARNING << "failed to get device properties for device " << didx
-                  << ", GPU metrics will not be available for this device: "
-                  << cudaGetErrorString(cudaerr);
-      continue;
-    }
+  dcgmerr = dcgmUpdateAllFields(dcgm_handle_, 1);
+  if (dcgmerr != DCGM_ST_OK) {
+    LOG_WARNING << "DCGM failed to update all fields, GPU metrics will not "
+                   "be available: "
+                << errorString(dcgmerr);
```
Comment (deadeyegoodwin) [inline src/core/metrics.cc:190]: Follow consistent pattern, report both message and number as above

File: src/core/metrics.cc [hunk 3]
```diff
@@ -156,205 +156,215 @@ Metrics::EnableGPUMetrics()
...
 
-  int dcnt;
-  cudaError_t cudaerr = cudaGetDeviceCount(&dcnt);
-  if (cudaerr != cudaSuccess) {
-    LOG_WARNING
-        << "failed to get device count, GPU metrics will not be available: "
-        << cudaGetErrorString(cudaerr);
+  dcgmerr = dcgmStartEmbedded(DCGM_OPERATION_MODE_MANUAL, &dcgm_handle_);
+  if (dcgmerr != DCGM_ST_OK) {
+    LOG_WARNING << "Error: dcgmStartEmbedded returned \""
```
Comment (deadeyegoodwin) [inline src/core/metrics.cc:181]: Follow lowercase pattern... "error, dcgm..."

File: src/core/metrics.cc [hunk 4]
```diff
@@ -156,205 +156,215 @@ Metrics::EnableGPUMetrics()
...
-      LOG_WARNING << "failed to get Bus ID for device " << didx
-                  << ", GPU metrics will not be available for this device: "
-                  << cudaGetErrorString(cudaerr);
-      continue;
-    }
+  unsigned int all_gpu_ids[DCGM_MAX_NUM_DEVICES];
+  int gpu_count;
+  dcgmerr = dcgmGetAllDevices(dcgm_handle_, all_gpu_ids, &gpu_count);
+  if (dcgmerr != DCGM_ST_OK) {
+    LOG_WARNING << "failed to get device info and count, GPU metrics will not "
```
Comment (deadeyegoodwin) [inline src/core/metrics.cc:198]: "error, unable to get ..."

File: src/core/metrics.cc [hunk 5]
```diff
@@ -156,205 +156,215 @@ Metrics::EnableGPUMetrics()
...
-        cudaDeviceGetPCIBusId(pcibusid_str, sizeof(pcibusid_str) - 1, didx);
-    if (cudaerr != cudaSuccess) {
-      LOG_WARNING << "failed to get Bus ID for device " << didx
-                  << ", GPU metrics will not be available for this device: "
-                  << cudaGetErrorString(cudaerr);
-      continue;
-    }
+  unsigned int all_gpu_ids[DCGM_MAX_NUM_DEVICES];
+  int gpu_count;
+  dcgmerr = dcgmGetAllDevices(dcgm_handle_, all_gpu_ids, &gpu_count);
```
Comment (deadeyegoodwin) [inline src/core/metrics.cc:196]: Does DCGM only return the cuda visible devices so the filtering is not needed?

File: src/core/metrics.cc [hunk 6]
```diff
@@ -156,205 +156,215 @@ Metrics::EnableGPUMetrics()
...
+                << errorString(dcgmerr);
+    return false;
+  }
 
-    nvmlDevice_t gpu;
-    nvmlReturn_t nvmlerr = nvmlDeviceGetHandleByPciBusId_v2(pcibusid_str, &gpu);
-    if (nvmlerr != NVML_SUCCESS) {
-      LOG_WARNING << "failed to get device from Bus ID "
+  // Get DCGM metrics for each GPU. Some devices may have problems using DCGM
+  // API and thus device count/ids needs to be updated.
```
Comment (deadeyegoodwin) [inline src/core/metrics.cc:205]: I don't think this comment is quite right... we are just collecting the gpus that support dcgm and can only report metrics for those gpus

File: src/core/metrics.cc [hunk 7]
```diff
@@ -156,205 +156,215 @@ Metrics::EnableGPUMetrics()
...
+  std::vector<uint32_t> available_gpu_ids;
+  dcgmDeviceAttributes_t gpu_attributes[DCGM_MAX_NUM_DEVICES];
+  for (int i = 0; i < gpu_count; i++) {
+    gpu_attributes[i].version = dcgmDeviceAttributes_version;
+    dcgmerr = dcgmGetDeviceAttributes(
+        dcgm_handle_, all_gpu_ids[i], &gpu_attributes[i]);
+    if (dcgmerr != DCGM_ST_OK) {
+      LOG_WARNING << "failed to get device properties for device "
```
Comment (deadeyegoodwin) [inline src/core/metrics.cc:213]: failed -> error

File: src/core/metrics.cc [hunk 8]
```diff
@@ -156,205 +156,215 @@ Metrics::EnableGPUMetrics()
...
+          dcgmReturn_t dcgmerr = dcgmGetLatestValuesForFields(
+              handle, available_gpu_ids[didx], fields, field_count,
+              field_values);
+          if (dcgmerr != DCGM_ST_OK) {
+            power_limit_fail_cnt[didx]++;
+            power_usage_fail_cnt[didx]++;
+            energy_fail_cnt[didx]++;
+            util_fail_cnt[didx]++;
+            mem_fail_cnt[didx]++;
+            LOG_WARNING << "failed to get field values for device "
```
Comment (deadeyegoodwin) [inline src/core/metrics.cc:280]: failed -> error

File: src/core/metrics.cc [hunk 9]
```diff
@@ -156,205 +156,215 @@ Metrics::EnableGPUMetrics()
...
+            // Power limit
+            if (power_limit_fail_cnt[didx] < fail_threshold) {
+              unsigned int power_limit = field_values[0].value.i64;
+              if ((field_values[0].status == DCGM_ST_OK) &&
+                  (!DCGM_INT64_IS_BLANK(power_limit))) {
+                power_limit_fail_cnt[didx] = 0;
+              } else {
+                power_limit_fail_cnt[didx]++;
+                power_limit = 0;
+                LOG_WARNING << "failed to get power limit for GPU " << didx
```
Comment (deadeyegoodwin) [inline src/core/metrics.cc:293]: failed -> error

File: src/test/custom/identity/CMakeLists.txt [hunk 10]
```diff
@@ -52,7 +52,7 @@ if(${TRITON_ENABLE_GPU})
 target_include_directories(identity PRIVATE ${CUDA_INCLUDE_DIRS})
 target_link_libraries(
   identity
-  PUBLIC ${CUDA_nvml_LIBRARY}
```
Comment (deadeyegoodwin) [inline src/test/custom/identity/CMakeLists.txt:55]: Does this really need to link against DCGM (or previously NVML)?

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

File: src/core/metrics.cc (hunk 1 above)
Comment (CoderHam) [reply src/core/metrics.cc:175]: Yes that is correct

File: src/core/metrics.cc (hunk 6 above)
Comment (CoderHam) [reply src/core/metrics.cc:205]: Fixed to -> "... thus these devices needs to be ignored."

File: src/core/metrics.cc (hunk 5 above)
Comment (CoderHam) [reply src/core/metrics.cc:194]: Yes, DCGM will report a list of all visible devices but it is possible dcgmGetDeviceAttributes is unable to get the properties for some of those devices

File: src/test/custom/identity/CMakeLists.txt (hunk 10 above)
Comment (CoderHam) [reply src/test/custom/identity/CMakeLists.txt]: No, reverted to NVML

File: src/core/metrics.cc (hunk 5 above)
Comment (deadeyegoodwin) [reply src/core/metrics.cc:194]: To be clear, by visible devices I mean only those visisble by CUDA due to CUDA_VISIBLE_DEVICES (and NVIDIA_VISIBLE_DEVICES in containers)

File: src/test/custom/identity/CMakeLists.txt (hunk 10 above)
Comment (deadeyegoodwin) [reply src/test/custom/identity/CMakeLists.txt]: You shouldn't need to link either. We no longer use nvml at all, and I don't know why the legacy custom backend was linking against it in the first place...

**Comment by CoderHam** [comment]
Closing for new #2924

**Pull request closed without merging**